jax/_src/scipy/stats/vonmises.py

```diff
@@ -0,0 +1,38 @@
+# Copyright 2022 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import scipy.stats as osp_stats
+from jax import lax
+from jax._src.lax.lax import _const as _lax_const
+from jax._src.numpy import lax_numpy as jnp
+from jax._src.numpy.util import _wraps, _promote_args_inexact
+from jax._src.typing import Array, ArrayLike
+
+def _cosm1(x: ArrayLike) -> Array:
+  minus_one = _lax_const(x, -1)
+  return lax.add(lax.cos(x), minus_one)
+
+def _pdf(x: ArrayLike, kappa: ArrayLike) -> Array:
+  return lax.exp(kappa * _cosm1(x)) / (2 * jnp.pi * lax.bessel_i0e(kappa))
+
+@_wraps(osp_stats.vonmises.pdf, update_doc=False)
+def pdf(x: ArrayLike, kappa: ArrayLike) -> Array:
+  x, kappa = _promote_args_inexact('vonmises.pdf', x, kappa)
+  zero = _lax_const(kappa, 0)
+  return jnp.where(lax.gt(kappa, zero), _pdf(x, kappa), jnp.nan)
+
+@_wraps(osp_stats.vonmises.logpdf, update_doc=False)
+def logpdf(x: ArrayLike, kappa: ArrayLike) -> Array:
+  return lax.log(pdf(x, kappa))
```

Comment:
I've made the other updates, but I'm not sure what you mean here. Are you referring to the first term of the pdf calculation? I'm not sure where the exponent of the logpdf is being taken.

Reply:
In the `_pdf` function, you compute `kappa * (jnp.cos(x) - 1) ) / (2 * jnp.pi * lax.bessel_i0e(kappa)`, which is essentially `logpdf`. You then pass it to `lax.exp` in order to compute the pdf, then here pass that result to `lax.log`, so the return value here is essentially `log(exp(logpdf))`. Because floating point arithmetic is not exact, it would be better to avoid taking the log of the exponent of the result that you want to return. Does that make sense?

Reply:
You can see this pattern used for other distributions, for example here: https://github.com/google/jax/blob/08b6b0dd43674da97ca00b957d16488845e7931f/jax/_src/scipy/stats/norm.py#L28-L39

Reply:
`lax.exp` doesn't completely wrap the entirety of `kappa * (jnp.cos(x) - 1) ) / (2 * jnp.pi * lax.bessel_i0e(kappa)` - only `kappa * (jnp.cos(x) - 1)` is passed to it. It's part of the actual pdf calculation on all of the examples I can see on the internet, so I don't think it's computing the exponent of the logpdf

i.e. [link](https://mathworld.wolfram.com/vonMisesDistribution.html)

Reply:
I see, sorry I misread the computation.

Still, I think we need to change this to compute logpdf directly, and then implement `pdf` by computing `exp(logpdf)`. The reason `logpdf` exists at all is because computing `logpdf` using `log(pdf)` in general loses accuracy for small probabilities, and accurate logs of small probabilities are important for many applications.

To be more explicit, I think the code should look something like this (untested):
```python
@_wraps(osp_stats.vonmises.pdf, update_doc=False)
def pdf(x: ArrayLike, kappa: ArrayLike) -> Array:
  return jnp.exp(logpdf(x, kappa))

@_wraps(osp_stats.vonmises.logpdf, update_doc=False)
def logpdf(x: ArrayLike, kappa: ArrayLike) -> Array:
  x, kappa = _promote_args_inexact('vonmises.logpdf', x, kappa)
  return jnp.where(kappa > 0, kappa * (jnp.cos(x) - 1) - jnp.log(2 * jnp.pi * lax.bessel_i0e(kappa)), jnp.nan)
```

Reply:
This makes sense. Altered, but not sure why the docs aren't building.